csound-plugins: init at 1.0.2

rvl opened the pull request:
## Description of changes

Some features which were previously built in to Csound have been moved to an [external plugins repository](https://github.com/csound/plugins).

Examples of removed functionality includes the Jacko opcodes, FLTK, and the virtual MIDI keyboard.

This patch allows the lost opcodes to be used again.

## Things done

- [x] Built on platform(s): `x86_64-linux`, `x86_64-linux.pkgsCross.aarch64-multiplatform`, `x86_64-linux.pkgsCross.raspberryPi`
- [x] Tested: has a package test to check for plugin opcodes
- [x] Tested basic functionality of all binary files
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

AndersonTorres commented on pkgs/applications/audio/csound/plugins.nix:
```diff
@@ -0,0 +1,78 @@
+{ lib, stdenv, fetchFromGitHub, cmake, gettext, runCommand
+, csound
+, libjack2 ? null
```
Never use null as default. It can have nasty consequences.

If you need to control what to include or exclude, use guard flags, like

```suggestion
, enableLibjack2 ? true
, libjack2
```

rvl replied:
What are the nasty consequences?

rvl replied:
I have added `enableX` flags for each plugin.

AndersonTorres replied:
This is hard to explain. A concrete example is better:

https://github.com/NixOS/nixpkgs/pull/233301#issuecomment-1567661913

AndersonTorres commented on pkgs/top-level/all-packages.nix:
```diff
@@ -30116,6 +30116,10 @@ with pkgs;
     inherit (pkgs.darwin.apple_sdk.frameworks) Accelerate AudioUnit CoreAudio CoreMIDI;
   };
 
+  csound-plugins = callPackage ../applications/audio/csound/plugins.nix {
```
Given that by-name is mandatory for new packages, please migrate it.

Hint: use passthru.

rvl replied:
OK, will do.

rvl replied:
Done.

AndersonTorres commented on pkgs/applications/audio/csound/default.nix:
```diff
@@ -30,20 +25,18 @@ stdenv.mkDerivation rec {
-    ++ lib.optional stdenv.isDarwin "-DCS_FRAMEWORK_DEST=${placeholder "out"}/lib"
+  cmakeFlags = lib.optional stdenv.isDarwin "-DCS_FRAMEWORK_DEST=${placeholder "out"}/lib"
     # Ignore gettext in CMAKE_PREFIX_PATH on cross to prevent find_program picking up the wrong gettext
     ++ lib.optional (stdenv.hostPlatform != stdenv.buildPlatform) "-DCMAKE_IGNORE_PATH=${lib.getBin gettext}/bin";
```
Can you use `canExecute` instead?

rvl replied:
OK, will have a look, thanks.

rvl replied:
`!(stdenv.buildPlatform.canExecute stdenv.hostPlatform)` has a slightly different meaning to `stdenv.hostPlatform != stdenv.buildPlatform`. I don't want to change this and break something.

@lilyinstarlight added it in #227135.

I have tried building `.#legacyPackages.x86_64-linux.pkgsCross.aarch64-multiplatform.{csound,csound-plugins}` to test it. Both can build with that cmake flag completely removed.

rvl committed:
csound: Fix build on raspberryPi

rvl committed:
csound-plugins: init at 1.0.2

Some features which were previously built in to Csound have been moved
to an external plugins repository.

Examples of removed functionality includes the Jacko opcodes, FLTK,
and the virtual MIDI keyboard.

This patch allows the lost opcodes to be used again.

rvl commented:
@AndersonTorres, thanks for the review - I have addressed your comments. Can you please take another look?

AndersonTorres commented:
Well, the csound expression has too many `pkg ? null`. This is a dangerous anti-pattern, and I would suggest to get rid of it.

bue it is better to do this in another PR.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/4125
